umap/ws.py

```diff
@@ -0,0 +1,98 @@
+#!/usr/bin/env python
+
+import asyncio
+from collections import defaultdict
+from typing import Literal, Optional
+
+import django
+import websockets
+from django.conf import settings
+from django.core.signing import TimestampSigner
+from pydantic import BaseModel, ValidationError
+from websockets import WebSocketClientProtocol
+from websockets.server import serve
+
+# This needs to run before the django-specific imports
+# See https://docs.djangoproject.com/en/5.0/topics/settings/#calling-django-setup-is-required-for-standalone-django-usage
+from umap.settings import settings_as_dict
+
+settings.configure(**settings_as_dict)
+django.setup()
+
+from umap.models import Map, User  # NOQA
+
+# Contains the list of websocket connections handled by this process.
+# It's a mapping of map_id to a set of the active websocket connections
+CONNECTIONS = defaultdict(set)
+
+
+class JoinMessage(BaseModel):
+    kind: str = "join"
+    token: str
+
+
+class OperationMessage(BaseModel):
+    kind: str = "operation"
+    verb: str = Literal["upsert", "update", "delete"]
+    subject: str = Literal["map", "layer", "feature"]
+    metadata: Optional[dict] = None
+    key: Optional[str] = None
+
+
+async def join_and_listen(
+    map_id: int, permissions: list, user: str | int, websocket: WebSocketClientProtocol
+):
+    """Join a "room" whith other connected peers.
+
+    New messages will be broadcasted to other connected peers.
+    """
+    print(f"{user} joined room #{map_id}")
+    CONNECTIONS[map_id].add(websocket)
+    try:
+        async for raw_message in websocket:
+            # recompute the peers-list at the time of message-sending.
+            # as doing so beforehand would miss new connections
+            peers = CONNECTIONS[map_id] - {websocket}
+            # Only relay valid "operation" messages
+            try:
+                OperationMessage.model_validate_json(raw_message)
```

Comment:
It will be useful later, because some messages will not be relayed to all the other connected parties, so we will need to inspect the messages. A bit early, but I really believe it will be useful very soon.

Reply:
If it's about checking one key, we may do it explicitly ? Without the need and the new layer of pydantic. At list, pydantic seems overkill to me here.

Reply:
You can see pydantic as doing form validation. In our case, It's more than checking just one key, because each message type might have a different format, that we need to ensure (some messages will be for the server itself, for instance if we decide the server saves the status of the connected peers)

Reply:
It is true that Pydantic in the current code is not doing much.

In the future, as we will add other messages types, we will want to ensure the messages match a format expected by the server, for instance when the clients will send disconnect messages that we don't want to relay on the other parties, or when they will send messages for just one peer, which should not be broadcasted.

In these cases, Pydantic will prove really useful than now, as we will have to check complex messages, and ensure their format match what is expected by the server.

It is worth noting that it already is useful, as it makes sure that the `"join"` message is holding the proper `"token"` key, and that only `operation` messages are valid in the "room".

If it's not too much of a hassle, I would like to keep it in place to avoid doing back and forth between implementation types for the message validation logic.